### PR TITLE
Remove-Get concurrency fix

### PR DESCRIPTION
**Pull request by iulianpascalau**
- added unit test for the edge-case
- fixed the persister edge-case
- added & integrated a new type of cache that is removal-aware

### CODE REVIEW

File: leveldb/leveldbSerial.go [hunk 1]
```diff
@@ -262,6 +297,10 @@ func (s *SerialDB) putBatch() error {
 	result := <-ch
 	close(ch)
 
+	s.mutBatch.Lock()
```
Comment (iulianpascalau) [inline leveldb/leveldbSerial.go]: not a great solution, this can just set to nil the writing batch if another go-routine called putBatch and managed to set the writingBatch

Comment (iulianpascalau) [reply leveldb/leveldbSerial.go]: right, reverted

File: storageUnit/storageunit_test.go [hunk 2]
```diff
@@ -100,14 +129,47 @@ func TestPutPresentShouldOverwriteValue(t *testing.T) {
...
+func TestUnit_GetWithExplicitlyRemovedKeyShouldNotCallThePersister(t *testing.T) {
+	t.Parallel()
+
+	mdb := &testscommon.PersisterStub{
+		HasCalled: func(key []byte) error {
+			assert.Fail(t, "should have not called Has")
+			return nil
+		},
+		GetCalled: func(key []byte) ([]byte, error) {
+			assert.Fail(t, "should have not called Has")
```
Comment (sstanculeanu) [inline storageUnit/storageunit_test.go]:
```suggestion
			assert.Fail(t, "should have not called Get")
```

File: storageUnit/storageunit_test.go [hunk 3]
```diff
@@ -144,7 +210,36 @@ func TestHasNotPresent(t *testing.T) {
...
+func TestUnit_HasWithExplicitlyRemovedKeyShouldNotCallThePersister(t *testing.T) {
+	t.Parallel()
+
+	mdb := &testscommon.PersisterStub{
+		HasCalled: func(key []byte) error {
+			assert.Fail(t, "should have not called Has")
+			return nil
+		},
+		GetCalled: func(key []byte) ([]byte, error) {
+			assert.Fail(t, "should have not called Has")
```
Comment (sstanculeanu) [inline storageUnit/storageunit_test.go]:
```suggestion
			assert.Fail(t, "should have not called Get")
```

File: go.mod [hunk 4]
```diff
@@ -5,7 +5,7 @@ go 1.20
 require (
 	github.com/hashicorp/golang-lru v0.6.0
 	github.com/multiversx/concurrent-map v0.1.4
-	github.com/multiversx/mx-chain-core-go v1.2.16
+	github.com/multiversx/mx-chain-core-go v1.2.17-0.20230919104727-566f55d213ab
```
Comment (sstanculeanu) [inline go.mod]: don't forget proper tag

File: leveldb/batch.go [hunk 5]
```diff
@@ -69,6 +71,14 @@ func (b *batch) IsRemoved(key []byte) bool {
 	defer b.mutBatch.RUnlock()
 
 	_, found := b.removedData[string(key)]
+	if found {
+		return true
+	}
+	_, found = b.cachedData[string(key)]
```
Comment (SebastianMarian) [inline leveldb/batch.go:77]: Could be a situation when both, cachedData[key} and previouslyRemoved[key], return true?

File: rtcache/removalTrackingCache.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	if found {
+		return types.NotRemovedStatus
+	}
+
+	return types.UnknownRemovalStatus
+}
+
+// Clear is used to completely clear both caches.
+func (cache *removalTrackingCache) Clear() {
+	cache.mutCriticalArea.RLock()
```
Comment (SebastianMarian) [inline rtcache/removalTrackingCache.go]: Lock / Unlock ?

File: testscommon/cacherMock.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	return mock.Get(key)
+}
+
+// HasOrAdd -
+func (mock *cacherMock) HasOrAdd(key []byte, value interface{}, _ int) (has, added bool) {
+	mock.mut.Lock()
+	defer mock.mut.Unlock()
+
+	_, found := mock.data[string(key)]
+
```
Comment (SebastianMarian) [inline testscommon/cacherMock.go:73]: Missing here: if found { return found, !found}  ?

File: testscommon/cacherMock.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+	mock.data[string(key)] = value
+
+	return !found, found
```
Comment (SebastianMarian) [inline testscommon/cacherMock.go]: return found, !found ?

File: types/interface.go [hunk 9]
```diff
@@ -3,7 +3,19 @@ package types
...
+
+// RemovalStatus removal status defines the removal possibility of a key
+type RemovalStatus string
+
+const (
+	// UnknownRemovalStatus defines the unknown status
+	UnknownRemovalStatus RemovalStatus = "unknown"
+	// ExplicitlyRemovedStatus defines the explicitly removed status
+	ExplicitlyRemovedStatus RemovalStatus = "explicitly removed"
+	// NotRemovedStatus the key is not removed
```
Comment (SebastianMarian) [inline types/interface.go]: // NotRemovedStatus defines the not removed status ?

File: storageUnit/storageunit_test.go (hunk 2 above)
Comment (iulianpascalau) [reply storageUnit/storageunit_test.go]: fixed

File: storageUnit/storageunit_test.go (hunk 3 above)
Comment (iulianpascalau) [reply storageUnit/storageunit_test.go]: fixed

File: go.mod (hunk 4 above)
Comment (iulianpascalau) [reply go.mod]: after testing will do :+1:

File: leveldb/batch.go (hunk 5 above)
Comment (iulianpascalau) [reply leveldb/batch.go:77]: Yes. The situation is the following:
the key existed and a call to Delete was made. The batch was written, and the data was deleted from the persister
and the new batch was created, the previously removed data was passed to the new batch. Then, immediately we call Put with the same key. Now, both caches contain the same key.
All is fine because in this case, we want the `IsRemoved` function to return false. Added a test with this exact scenario.

File: rtcache/removalTrackingCache.go (hunk 6 above)
Comment (iulianpascalau) [reply rtcache/removalTrackingCache.go]: right, to preserve the consistency as much as we can

File: testscommon/cacherMock.go (hunk 7 above)
Comment (iulianpascalau) [reply testscommon/cacherMock.go:73]: right, changed also the other return as it was badly written

File: testscommon/cacherMock.go (hunk 8 above)
Comment (iulianpascalau) [reply testscommon/cacherMock.go]: right, thanks

File: types/interface.go (hunk 9 above)
Comment (iulianpascalau) [reply types/interface.go]: fixed

**Review by sstanculeanu — DISMISSED** [review]

File: testscommon/cacherMock.go [hunk 10]
```diff
@@ -67,10 +67,13 @@ func (mock *cacherMock) HasOrAdd(key []byte, value interface{}, _ int) (has, add
...
 
 	_, found := mock.data[string(key)]
+	if found {
+		return found, !found
+	}
 
 	mock.data[string(key)] = value
 
-	return !found, found
+	return found, !found
```
Comment (SebastianMarian) [inline testscommon/cacherMock.go:76]: :+1:

**Review by SebastianMarian — DISMISSED** [review]

File: leveldb/batch.go (hunk 5 above)
Comment (SebastianMarian) [reply leveldb/batch.go:77]: ok
